bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -0,0 +1,677 @@
+///
+/// Returning "None" will result in the C callback returning an error,
+/// canceling the connection.
+///
+/// See s2n_renegotiate_request_cb in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+//
+// This method returns Option instead of Result because the callback has no mechanism
+// for surfacing errors to the application, so Result would be somewhat deceptive.
+pub trait RenegotiateCallback: 'static + Send + Sync {
+    fn on_renegotiate_request(
+        &mut self,
+        connection: &mut Connection,
+    ) -> Option<RenegotiateResponse>;
+}
+
+impl RenegotiateCallback for RenegotiateResponse {
+    fn on_renegotiate_request(&mut self, _conn: &mut Connection) -> Option<RenegotiateResponse> {
+        Some(*self)
+    }
+}
+
+impl Connection {
+    /// Reset the connection so that it can be renegotiated.
+    ///
+    /// Returning "None" will result in the C callback returning an error,
+    /// canceling the connection.
+    ///
+    /// See s2n_renegotiate_wipe in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+    /// The Rust equivalent of the connection-specific methods listed are:
+    ///  - Methods to set the file descriptors: not currently supported by rust bindings
+    ///  - Methods to set the send callback:
+    ///    ([Connection::set_send_callback()], [Connection::set_send_context()])
+    ///  - Methods to set the recv callback:
+    ///    ([Connection::set_receive_callback()], [Connection::set_receive_context()])
+    /// Additionally, the application does not need to re-call:
+    /// - [Connection::set_application_context()]
+    /// - [Connection::set_waker()]
+    ///
+    fn wipe_for_renegotiate(&mut self) -> Result<(), Error> {
+        // Reset any values in the context that need to be reset
+        self.context_mut().connection_initialized = false;
+        unsafe {
+            // Retrieve the context before the wipe
+            let context = s2n_connection_get_ctx(self.as_ptr()).into_result()?;
+
+            s2n_renegotiate_wipe(self.as_ptr()).into_result()?;
+
+            // Re-set the context
+            s2n_connection_set_ctx(self.as_ptr(), context.as_ptr()).into_result()?;
+        }
+        Ok(())
+    }
+
+    /// Perform a new handshake on an already established connection.
+    ///
+    /// The first element of the returned pair represents progress on the new
+    /// handshake, like [Connection::poll_negotiate()].
+    ///
+    /// If any application data is received during the new handshake, the number
+    /// of bytes received is returned as the second element of the returned pair,
+    /// and the data is written to `buf`.
+    ///
+    /// See s2n_renegotiate in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+    fn poll_renegotiate(&mut self, buf: &mut [u8]) -> (Poll<Result<(), Error>>, usize) {
```

Comment:
No, the issue is that the `usize` is unrelated to the `Poll`. You can continue to receive application data during a renegotiated handshake.

The `Poll` tracks negotiate progress, but the `usize` is for application data received during the handshake. So negotiate progress can be `Pending`, but we read application data. If I do `Poll<Result<usize, Error>>`, you'd have to keep polling poll_renegotiate after `Ready`, since the handshake isn't done. Polling after `Ready` is undefined behavior.

I considered `(Poll<Result<(), Error>>, Poll<Result<usize, Error>>)` to explicitly track negotiate vs recv results, but the usize doesn't really need the `Poll` or the `Result`.